### PR TITLE
Comments: unlike actions to the data-layer

### PR DESCRIPTION
**Pull request by rodrigoi**
This PR adds the unlike action to the data layer, preserving current optimistic behavior.

to test, unlike a liked comment on the post list or the reader. To run unit tests:

```
npm run test-client client/state/comments
```

```sh
npm run test-client client/state/data-layer/wpcom/sites/comments/likes
```

depends on #15360

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleLikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic updated on error
+	dispatch( { type: COMMENTS_LIKE, siteId, postId, commentId } );
+	// dispatch a error notice
+	dispatch( errorNotice( translate( 'Could not unlike this comment' ) ) );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Unlike this comment.' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleUnlikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic updated on error
+	dispatch( { type: COMMENTS_LIKE, siteId, postId, commentId } );
+	// dispatch a error notice
+	dispatch( errorNotice( translate( 'Could not unlike this comment' ) ) );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Unlike this comment.' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by dmsnell** [comment]
What's the point of the `COMMENTS_LIKE_UPDATE` action? How is it different than `COMMENTS_LIKE` and `COMMENT_UNLIKE`?

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		path: `/sites/${ action.siteId }/comments/${ action.commentId }/likes/mine/delete`
+	}, action ) );
+};
+
+export const updateCommentLikes = ( { dispatch }, { siteId, postId, commentId }, next, { i_like, like_count } ) => dispatch( local( {
+	type: COMMENTS_LIKE,
+	siteId,
+	postId,
+	commentId,
+	i_like,
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: are we going to update the reducer here because it looks like it's not watching for `i_like`, only passing `true` for that value. further, if we wanted to clear this up a bit we could choose the action based on `i_like`…

```js
{
	type: i_like ? COMMENTS_LIKE : COMMENS_UNLIKE,
	siteId,
	postId,
	commentId,
	like_count,
}
```

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { local } from 'state/data-layer/utils';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { errorNotice } from 'state/notices/actions';
+
+export const unlikeComment = ( { dispatch }, action ) => {
+	dispatch( http( {
+		method: 'POST',
+		apiVersion: '1.1',
+		path: `/sites/${ action.siteId }/comments/${ action.commentId }/likes/mine/delete`
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: so many endpoints…so sad…

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		unlikeComment( { dispatch }, action );
+
+		expect( dispatch ).to.have.been.calledOnce;
+		expect( dispatch ).to.have.been.calledWith( http( {
+			apiVersion: '1.1',
+			method: 'POST',
+			path: `/sites/${ SITE_ID }/comments/1/likes/mine/delete`,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js]: I'm guessing there's a reason to repeat `action` here instead of passing it as the second param to `http()`? or was it just a slip?

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+		updateCommentLikes( { dispatch }, { siteId: SITE_ID, postId: POST_ID, commentId: 1 }, null, { i_like: true, like_count: 4 } );
+
+		expect( dispatch ).to.have.been.calledOnce;
+		expect( dispatch ).to.have.been.calledWith( {
+			type: COMMENTS_LIKE,
+			siteId: SITE_ID,
+			postId: POST_ID,
+			commentId: 1,
+			iLike: true,
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js]: on an earlier note I don't think we need to carry the `iLike` here because it seems already captured in the action type. if we wanted to send a like with `iLike: false` then I think it would be more appropriate to send `COMMENTS_UNLIKE`

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	}, action ) );
+};
+
+export const updateCommentLikes = ( { dispatch }, { siteId, postId, commentId }, next, { i_like, like_count } ) => dispatch( local( {
+	type: COMMENTS_LIKE,
+	siteId,
+	postId,
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: does the post id get used? is it even relevant? I know this is a bit murky since WP ids comments by site and comment id while still preserving a link to the original post. it didn't seem like that was being used here.

maybe it's worth passing even if it's not used…

**Comment by dmsnell** [comment]
@rodrigoi I left a few questions on the data being passed around. I think we can merge this soon, but I'd like to clarify the action types in use. thanks for staying diligent on it!

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js (hunk 4 above)
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: 😢

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+export const unlikeComment = ( { dispatch }, action ) => {
+	dispatch( http( {
+		method: 'POST',
+		apiVersion: '1.1',
+		path: `/sites/${ action.siteId }/comments/${ action.commentId }/likes/mine/delete`
+	}, action ) );
+};
+
+export const updateCommentLikes = ( { dispatch }, { siteId, postId, commentId }, next, { i_like, like_count } ) => dispatch( local( {
+	type: COMMENTS_LIKE,
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: So the reducer does listen to `COMMENTS_UNLIKE`, I'm a little confused as to why we're dispatching this one. Is it supposed to update comment count?

**Comment by gwwar** [comment]
@rodrigoi do you mind rebasing?

**Comment by rodrigoi** [comment]
@dmsnell rebased, updated and tested 😉 . Let me know what you think.

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+} from 'state/action-types';
+import {
+	unlikeComment,
+	updateCommentLikes,
+	handleUnlikeFailure,
+} from '../';
+import { local } from 'state/data-layer/utils';
+import { http } from 'state/data-layer/wpcom-http/actions';
+
+const SITE_ID = 91750058;
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js]: Do you mind updating this to `77203074`?

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+} ) );
+
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleUnlikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic updated on error
+	dispatch( { type: COMMENTS_LIKE, siteId, postId, commentId } );
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: Should this be a local action, or did we want to dispatch a network request?

Comment (rodrigoi) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: good catch. We should make those local. Fixed on 7e6d97f2afc79880c7a984d77f6455fbcbbcbc9a.

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+export const unlikeComment = ( { dispatch }, action ) => {
+	dispatch( http( {
+		method: 'POST',
+		apiVersion: '1.1',
+		path: `/sites/${ action.siteId }/comments/${ action.commentId }/likes/mine/delete`
+	}, action ) );
+};
+
+export const updateCommentLikes = ( { dispatch }, { siteId, postId, commentId }, next, { like_count } ) => dispatch( local( {
+	type: COMMENTS_UNLIKE,
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: We can leave this for another PR, but I feel like this would be clearer with an explicit action. For example `COMMENT_COUNT_SET`, if we periodically want to sync the count from network responses.

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleUnlikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic updated on error
+	dispatch( local( { type: COMMENTS_LIKE, siteId, postId, commentId } ) );
+	// dispatch a error notice
+	dispatch( errorNotice( translate( 'Could not unlike this comment' ) ) );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Unlike this comment.' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/comments/reducer.js [hunk 13]
```diff
@@ -78,21 +77,11 @@ export function items( state = {}, action ) {
...
-			return {
-				...state,
-				[ stateKey ]: map( state[ stateKey ], updateComment( commentId, { i_like: iLike, like_count: likeCount } ) )
-			};
 		case COMMENTS_UNLIKE:
+			const { like_count } = action;
 			return {
 				...state,
-				[ stateKey ]: map( state[ stateKey ], updateComment( commentId, { i_like: false } ) )
+				[ stateKey ]: map( state[ stateKey ], updateComment( commentId, { i_like: type === COMMENTS_LIKE, like_count } ) )
```
Comment (dmsnell) [inline client/state/comments/reducer.js]: I see the value in the fall-through logic here, but we generally like to avoid it. if it's only saving us a line or two we might consider just leaving the two branches separate. the logic on `i_like`, for example, is a bit confusing to me when I see it right under `COMMENTS_UNLIKE`

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+	like_count
+} ) );
+
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleUnlikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic updated on error
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: quick typo: `updated` → `update`

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js (hunk 9 above)
Comment (rodrigoi) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/test/index.js]: almost missed this comment... changed on 839e5e1438a366a9236967bc39faf508e48e2cdc.

File: client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+/***
+ * dispatches a error notice if creating a new comment request failed
+ *
+ * @param {Function} dispatch redux dispatcher
+ */
+export const handleUnlikeFailure = ( { dispatch }, { siteId, postId, commentId } ) => {
+	// revert optimistic update on error
+	dispatch( local( { type: COMMENTS_LIKE, siteId, postId, commentId } ) );
+	// dispatch a error notice
+	dispatch( errorNotice( translate( 'Could not unlike this comment' ) ) );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/comments/likes/mine/delete/index.js]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Unlike this comment.' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/comments/reducer.js (hunk 13 above)
Comment (rodrigoi) [inline client/state/comments/reducer.js]: confusion removed on 0e0b46a559787ce4a8dbb8f14cb84ca4423440c6 😉
